pkg/auth/tokenexchange/middleware.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	if config.HeaderStrategy != "" &&
+		config.HeaderStrategy != HeaderStrategyReplace &&
+		config.HeaderStrategy != HeaderStrategyCustom {
+		return fmt.Errorf("invalid header_strategy: %s (valid values: '%s', '%s')",
+			config.HeaderStrategy, HeaderStrategyReplace, HeaderStrategyCustom)
+	}
+
+	return nil
+}
+
+// injectToken handles token injection based on the configured strategy
+func injectToken(r *http.Request, token string, config Config) error {
+	strategy := config.HeaderStrategy
+	if strategy == "" {
+		strategy = HeaderStrategyReplace // Default to replace for backwards compatibility
+	}
+
+	switch strategy {
+	case HeaderStrategyReplace:
+		logger.Debugf("Token exchange successful, replacing Authorization header")
+		r.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+		return nil
+
+	case HeaderStrategyCustom:
+		if config.ExternalTokenHeaderName == "" {
+			return fmt.Errorf("external_token_header_name must be specified when header_strategy is '%s'", HeaderStrategyCustom)
+		}
+		logger.Debugf("Token exchange successful, adding token to custom header: %s", config.ExternalTokenHeaderName)
+		r.Header.Set(config.ExternalTokenHeaderName, fmt.Sprintf("Bearer %s", token))
+		return nil
+
+	default:
+		return fmt.Errorf("unsupported header_strategy: %s (valid values: '%s', '%s')",
+			strategy, HeaderStrategyReplace, HeaderStrategyCustom)
+	}
+}
+
+// CreateTokenExchangeMiddlewareFromClaims creates a middleware that uses token claims
+// from the auth middleware to perform token exchange.
+// This is a public function for direct usage in proxy commands.
+func CreateTokenExchangeMiddlewareFromClaims(config Config) types.MiddlewareFunction {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Get claims from the auth middleware
+			claims, ok := r.Context().Value(auth.ClaimsContextKey{}).(jwt.MapClaims)
+			if !ok {
+				logger.Debug("No claims found in context, proceeding without token exchange")
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			// Extract the original token from the Authorization header
+			authHeader := r.Header.Get("Authorization")
+			if authHeader == "" || !strings.HasPrefix(authHeader, "Bearer ") {
+				logger.Debug("No valid Bearer token found, proceeding without token exchange")
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			subjectToken := strings.TrimPrefix(authHeader, "Bearer ")
+			if subjectToken == "" {
+				logger.Debug("Empty Bearer token, proceeding without token exchange")
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			// Log some claim information for debugging
+			if sub, exists := claims["sub"]; exists {
+				logger.Debugf("Performing token exchange for subject: %v", sub)
+			}
+
+			// Build scopes array
+			scopes := []string{}
+			if config.Scope != "" {
+				scopes = strings.Split(config.Scope, " ")
```

Comment:
scopes are always separated by whitespaces?

Reply:
so these are scopes the user will set to be requested during the exchange..fair comment that we should probably normalize them sooner (in the CLI perhaps) and pass around a slice from the upper layer

Reply:
I ended up accepting a `[]string` in the config struct I agree with Yolanda that the middleware `Config` structure shouldn't depend on a particular way of formatting scopes (the space-separated format is something that we send over the wire because that's what the RFC mandates)